pkgs/development/compilers/openjdk/generic.nix

```diff
@@ -0,0 +1,664 @@
+      ${lib.optionalString (!headless) ''
+        mkdir -p $jre/lib/openjdk/jre/lib/fonts
+        ln -s ${liberation_ttf}/share/fonts/truetype $jre/lib/openjdk/jre/lib/fonts/fallback
+      ''}
+
+      # Remove duplicate binaries.
+      for i in $(cd $out/lib/openjdk/bin && echo *); do
+        if [ "$i" = java ]; then continue; fi
+        if cmp -s $out/lib/openjdk/bin/$i $jre/lib/openjdk/jre/bin/$i; then
+          ln -sfn $jre/lib/openjdk/jre/bin/$i $out/lib/openjdk/bin/$i
+        fi
+      done
+
+      # Generate certificates.
+      (
+        cd $jre/lib/openjdk/jre/lib/security
+        rm cacerts
+        perl ${./8/generate-cacerts.pl} $jre/lib/openjdk/jre/bin/keytool ${cacert}/etc/ssl/certs/ca-bundle.crt
+      )
+    ''
+    + ''
+
+      ln -s $out/lib/openjdk/bin $out/bin
+    ''
+    + lib.optionalString (!atLeast11) ''
+      ln -s $jre/lib/openjdk/jre/bin $jre/bin
+      ln -s $jre/lib/openjdk/jre $out/jre
+    '';
+
+  preFixup =
+    (
+      if atLeast11 then
+        ''
+          # Propagate the setJavaClassPath setup hook so that any package
+          # that depends on the JDK has $CLASSPATH set up properly.
+          mkdir -p $out/nix-support
+          #TODO or printWords?  cf https://github.com/NixOS/nixpkgs/pull/27427#issuecomment-317293040
+          echo -n "${setJavaClassPath}" > $out/nix-support/propagated-build-inputs
+        ''
+      else
+        ''
+          # Propagate the setJavaClassPath setup hook from the JRE so that
+          # any package that depends on the JRE has $CLASSPATH set up
+          # properly.
+          mkdir -p $jre/nix-support
+          printWords ${setJavaClassPath} > $jre/nix-support/propagated-build-inputs
+        ''
+    )
+    + ''
+
+      # Set JAVA_HOME automatically.
+      mkdir -p $out/nix-support
+      cat <<EOF > $out/nix-support/setup-hook
+      if [ -z "\''${JAVA_HOME-}" ]; then export JAVA_HOME=$out/lib/openjdk; fi
+      EOF
+    '';
+
+  postFixup = ''
+    # Build the set of output library directories to rpath against
+    LIBDIRS=""
+    for output in $(getAllOutputNames); do
+      if [ "$output" = debug ]; then continue; fi
+      LIBDIRS="$(find $(eval echo \$$output) -name \*.so\* -exec dirname {} \+ | ${
+        if atLeast17 then "sort -u" else "sort | uniq"
+      } | tr '\n' ':'):$LIBDIRS"
+    done
+    # Add the local library paths to remove dependencies on the bootstrap
+    for output in $(getAllOutputNames); do
+      if [ "$output" = debug ]; then continue; fi
+      OUTPUTDIR=$(eval echo \$$output)
+      BINLIBS=$(find $OUTPUTDIR/bin/ -type f; find $OUTPUTDIR -name \*.so\*)
+      echo "$BINLIBS" | while read i; do
+        patchelf --set-rpath "$LIBDIRS:$(patchelf --print-rpath "$i")" "$i" || true
+        patchelf --shrink-rpath "$i" || true
+      done
+    done
+  '';
+
+  # TODO: The OpenJDK 8 derivation got this wrong.
+  disallowedReferences = [
+    (if atLeast11 then jdk-bootstrap' else jdk-bootstrap)
+  ];
```

Comment:
See comment near top about the double override which makes this currently redundant.

Reply:
I changed it to be like this because the OpenJDK 8 derivation incorrectly lists the non‐overrided bootstrap JDK here, making it quite useless, but affecting the derivation hash. I’ve removed the override from the function header for now so that this can correctly match the original derivation and avoid the rebuilds.